### ISSUE TITLE
setuptools is removing easy_install, deprecation warning in 42.0.0

### ISSUE THREAD

**Issue by k4ml** [buildout/buildout#493]
Refer to pypa/setuptools/pull/1830.

Buildout still calling out to setuptools.command.easy_install in `zc/buildout/easy_install.py:call_easy_install(spec, dest)`.

So what our take and future direction on this?

42.0.0 changelog also mentioned about eggs are not supported anymore.

> #1830, #1909: Mark the easy_install script and setuptools command as deprecated, and use pip when available to fetch/build wheels for missing setup_requires/tests_require requirements, with the following differences in behavior:
> 
>         support for python_requires
>         better support for wheels (proper handling of priority with respect to PEP 425 tags)
>         PEP 517/518 support
>         eggs are not supported

**Comment by mauritsvanrees** [comment]
This also worries me. For now it seems to still work though.
I tried with Python 3.7.4 and this `requirements.txt`:

```
setuptools==42.0.1
zc.buildout==2.13.2
```

and this `buildout.cfg`:

```
[buildout]
parts = test
eggs-directory = eggs
download-cache = downloads
abi-tag-eggs = false

[test]
recipe = zc.recipe.egg
eggs = zest.releaser
```

Result is:

```
$ bin/buildout 
Creating directory '/Users/maurits/tmp/s42/downloads'.
Creating directory '/Users/maurits/tmp/s42/eggs'.
Creating directory '/Users/maurits/tmp/s42/parts'.
Creating directory '/Users/maurits/tmp/s42/develop-eggs'.
Getting distribution for 'zc.recipe.egg>=2.0.6'.
WARNING: The easy_install command is deprecated and will be removed in a future version.
Got zc.recipe.egg 2.0.7.
Installing test.
Getting distribution for 'zest.releaser'.
WARNING: The easy_install command is deprecated and will be removed in a future version.
warning: no previously-included files found matching '.installed.cfg'
warning: no previously-included files found matching '.gitattributes'
warning: no previously-included files matching '*.pyc' found anywhere in distribution
Got zest.releaser 6.19.1.
Getting distribution for 'twine>=1.6.0'.
Got twine 3.1.0.
Getting distribution for 'six'.
Got six 1.13.0.
...
Generated script '/Users/maurits/tmp/s42/bin/release'.
```

The generated scripts seem to work.
The `eggs` directory contains 23 eggs.
The `downloads` directory contains 21 wheels and two tar balls.

The tar balls are for `zc.recipe.egg` and `zest.releaser`, which are the ones that trigger the `WARNING` above. When I pin `zest.releaser = 6.19.0` in the versions, I don't get the warning, because I did that release, and apparently I have the `wheel` package installed and @reinout not. :-)

So:
- If all your packages are available in the wheel format, you should still be fine. On install, this will be turned into an egg without problems.
- Tar balls (or zip files) still work, but with a warning.
- I am not sure about packages for which (only) an egg has been uploaded to PyPI.

**Comment by k4ml** [comment]
It will still fail if you remove `call_easy_install()` function in zc/buildout/easy_install.py:-

```
 ./bin/buildout 
Unused options for buildout: 'unzip'.
Installing base.
Getting distribution for 'psycopg2-binary'.
While:
  Installing base.
  Getting distribution for 'psycopg2-binary'.

An internal error occurred due to a bug in either zc.buildout or in a
recipe being used:
Traceback (most recent call last):
  File "/home/kamal/git/xbuildout/.env/lib/python3.6/site-packages/zc/buildout/buildout.py", line 2174, in main
    getattr(buildout, command)(args)
  File "/home/kamal/git/xbuildout/.env/lib/python3.6/site-packages/zc/buildout/buildout.py", line 817, in install
    installed_files = self[part]._call(recipe.install)
  File "/home/kamal/git/xbuildout/.env/lib/python3.6/site-packages/zc/buildout/buildout.py", line 1603, in _call
    return f()
  File "/home/kamal/git/xbuildout/eggs/zc.recipe.egg-2.0.7-py3.6.egg/zc/recipe/egg/egg.py", line 227, in install
    reqs, ws = self.working_set()
  File "/home/kamal/git/xbuildout/eggs/zc.recipe.egg-2.0.7-py3.6.egg/zc/recipe/egg/egg.py", line 87, in working_set
    allow_unknown_extras=bool_option(buildout_section, 'allow-unknown-extras')
  File "/home/kamal/git/xbuildout/eggs/zc.recipe.egg-2.0.7-py3.6.egg/zc/recipe/egg/egg.py", line 168, in _working_set
    allow_unknown_extras=allow_unknown_extras)
  File "/home/kamal/git/xbuildout/.env/lib/python3.6/site-packages/zc/buildout/easy_install.py", line 957, in install
    return installer.install(specs, working_set)
  File "/home/kamal/git/xbuildout/.env/lib/python3.6/site-packages/zc/buildout/easy_install.py", line 682, in install
    for dist in self._get_dist(requirement, ws):
  File "/home/kamal/git/xbuildout/.env/lib/python3.6/site-packages/zc/buildout/easy_install.py", line 574, in _get_dist
    dists = [_move_to_eggs_dir_and_compile(dist, self._dest)]
  File "/home/kamal/git/xbuildout/.env/lib/python3.6/site-packages/zc/buildout/easy_install.py", line 1744, in _move_to_eggs_dir_and_comp
ile
    unpacker = UNPACKERS.get(ext, call_easy_install)
NameError: name 'call_easy_install' is not defined
```
But fortunately that function is just referenced and not get called, so we can just stub it as:-

```
def call_easy_install(spec, dest): pass
```
And installing wheel will still work.

**Comment by k4ml** [comment]
But it will fail to install package that doesn't has wheel such as https://pypi.org/project/pyspf/#files.

**Comment by gotcha** [comment]
@mauritsvanrees Would it make sense to use the buildout you describe (or sthing similar) as part of integration/end-to-end testing of buildout ?

**Comment by gotcha** [comment]
@k4ml Can you post an example buildout with `pyspf` ?

**Comment by k4ml** [comment]
@gotcha 

```
[buildout]
parts = base
unzip = true
newest = false

[base]
recipe = zc.recipe.egg
eggs = 
    pyspf
interpreter = python
```

**Comment by gotcha** [comment]
@k4ml Is this supposed to succeed or fail ?

**Comment by gotcha** [comment]
@mauritsvanrees @k4ml I have setup Actions on a buildout/buildout fork: https://github.com/gotcha/buildout/actions?query=workflow%3A%22Test+generating+scripts%22

Is that useful ? Should that be merged ?

**Comment by mauritsvanrees** [comment]
The `pyspf` example fails for me, but it already fails with earlier setuptools 41.6.0.
With 42.0.1 the traceback is:

```
Installing base.
While:
  Installing base.

An internal error occurred due to a bug in either zc.buildout or in a
recipe being used:
Traceback (most recent call last):
  File "/Users/maurits/tmp/s42/lib/python3.7/site-packages/zc/buildout/buildout.py", line 2174, in main
    getattr(buildout, command)(args)
  File "/Users/maurits/tmp/s42/lib/python3.7/site-packages/zc/buildout/buildout.py", line 817, in install
    installed_files = self[part]._call(recipe.install)
  File "/Users/maurits/tmp/s42/lib/python3.7/site-packages/zc/buildout/buildout.py", line 1603, in _call
    return f()
  File "/Users/maurits/tmp/s42/eggs/zc.recipe.egg-2.0.7-py3.7.egg/zc/recipe/egg/egg.py", line 263, in install
    relative_paths=self._relative_paths,
  File "/Users/maurits/tmp/s42/lib/python3.7/site-packages/zc/buildout/easy_install.py", line 1225, in scripts
    generated.extend(_pyscript(spath, sname, rpsetup, initialization))
  File "/Users/maurits/tmp/s42/lib/python3.7/site-packages/zc/buildout/easy_install.py", line 1447, in _pyscript
    changed = _file_changed(dest, contents)
  File "/Users/maurits/tmp/s42/lib/python3.7/site-packages/zc/buildout/easy_install.py", line 1351, in _file_changed
    return f.read() != old_contents
  File "/usr/local/bin/../Cellar/python/3.7.4_1/bin/../Frameworks/Python.framework/Versions/3.7/lib/python3.7/codecs.py", line 322, in decode
    (result, consumed) = self._buffer_decode(data, self.errors, final)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xcf in position 0: invalid continuation byte
```

**Comment by mauritsvanrees** [comment]
@gotcha I have not used GitHub actions before. Maybe the same could be done with Travis. But this could be useful yes.
I have not deeply checked your code, but some smoke test like this would be useful yes.

**Comment by k4ml** [comment]
> @k4ml Is this supposed to succeed or fail ?

@gotcha the buildout supposed to succeed but will fail if you stub out call_easy_install().

```
 ./bin/buildout -v
Installing 'zc.recipe.egg'.
Picked: zc.recipe.egg = 2.0.7
Unused options for buildout: 'unzip'.
Updating base.
Installing 'pyspf'.
We have no distributions for pyspf that satisfies 'pyspf'.
Getting distribution for 'pyspf'.
Fetching pyspf 2.0.13 from: https://files.pythonhosted.org/packages/7a/0c/1a10db5bf17e13c596d47b0a6903ac2837f78c2c8faf5918c748
cb83e3f8/pyspf-2.0.13.tar.gz#sha256=62dc1be39519e343202d480da7ef93d834b5a50cd4f217bef9335ed15292929b
Running easy_install:
"/home/kamal/git/xbuildout/.env/bin/python3.6" "-c" "import sys; sys.path[0:0] = ['/home/kamal/git/xbuildout/.env/lib/python3.
6/site-packages']; from setuptools.command.easy_install import main; main()" "-mZUNxd" "/home/kamal/git/xbuildout/eggs/tmpy62p
p65l" "-q" "/tmp/tmp_wa18am9get_dist/pyspf-2.0.13.tar.gz"
path=['/home/kamal/git/xbuildout/.env/lib/python3.6/site-packages']

WARNING: The easy_install command is deprecated and will be removed in a future version.
zip_safe flag not set; analyzing archive contents...
Got pyspf 2.0.13.
Picked: pyspf = 2.0.13
```
./bin/buildout --version
buildout version 2.13.2
```
```
>>> import setuptools
>>> setuptools.__version__
'42.0.1'
```

**Comment by k4ml** [comment]
@mauritsvanrees you can try this - https://github.com/k4ml/xbuildout. It works for me in both python 3.6 and 3.7.

**Comment by mauritsvanrees** [comment]
@k4ml Strange: your buildout works for me too, but I don't see a real difference with my own test buildout, which still fails. Maybe something on my end, or maybe Mac specific.

Anyway: your point is that such a package with only a tarball cannot be installed by current buildout when you basically remove `call_easy_install`. Buildout will not remove this call, but I think you mean that in the future this call may fail. So stubbing out the call gives us a glimpse of what would no longer work at that point.
It may not be entirely true: buildout has its own hacked up copy of `easy_install.py`, so `call_easy_install` might continue to work even when setuptools no longer has an easy_install. But yes: it is old code that may break at some point in the not too distant future.

**Comment by gotcha** [comment]
@mauritsvanrees This actually breaks tests: see #494

**Comment by k4ml** [comment]
> It may not be entirely true: buildout has its own hacked up copy of `easy_install.py`, so `call_easy_install` might continue to work even when setuptools no longer has an easy_install. But yes: it is old code that may break at some point in the not too distant future.

@mauritsvanrees zc/buildout/easy_install.py is not independent implementation of easy_install. Firstly it still import setuptools.command.easy_install. This direct import however seem to be used in just a single place in `_create_script()` function. But the heavy duty still being done by shelling out to easy_install command through subprocess.call(). You can see below what call_easy_install() actually doing:-

```
 ./bin/buildout 
Unused options for buildout: 'unzip'.
Updating base.
Getting distribution for 'pyspf'.
['/home/kamal/git/xbuildout/.env/bin/python3.6', '-c', "import sys; sys.path[0:0] = ['/home/kamal/git/xbuildout/.env/lib/python3.6/si
te-packages']; from setuptools.command.easy_install import main; main()", '-mZUNxd', '/home/kamal/git/xbuildout/eggs/tmpzeysyfmv', '-
q', '/tmp/tmp7r_3bumxget_dist/pyspf-2.0.13.tar.gz']
WARNING: The easy_install command is deprecated and will be removed in a future version.
zip_safe flag not set; analyzing archive contents...
Got pyspf 2.0.13.
```

So in the future, when setuptools finally removed easy_install, we have to vendorize them or use something like https://github.com/pypa/easy_install that @jaraco mentioned in pypa/setuptools/pull/1909.

**Comment by mauritsvanrees** [comment]
Ah, you are right.

**Comment by k4ml** [comment]
With a stubbed call_easy_install() function, it even failed early on trying to install zc.recipe.egg.

@mauritsvanrees I noticed you're working on plone. Since buildout heavily used in plone project, is there any discussion among plone devs on how to move forward with this?

**Comment by gotcha** [comment]
@k4ml I am also a Plonista, I guess you saw #494 and #495

What do you use buildout for ?

**Comment by k4ml** [comment]
@gotcha we use buildout to manage dependencies and config of our projects. Our dependencies basically a mix of packages on pypi, packages on github (zip tarball) and private packages on bitbucket (pulled through mr.developer). We also use it to manage our apache/wsgi config using our fork of collective.recipe.modwsgi.

I noticed #494 and #495 but how that directly related on future plan for easy_install?

**Comment by mauritsvanrees** [comment]
The direction I currently see for Plone, is:

- Create a virtualenv.
- Use pip to install all requirements, including buildout recipes and extensions.
- Call buildout to do the rest (mostly configuration management, and configure/make/make-install), but do not let it install any Python packages. Maybe a new buildout option `allow-python-package-install`, which by default would be true, but that we would set to false.

Requirements should be in `requirements.txt` and version constraints in `constraints.txt` for pip. I have a first version of a [script](https://gist.github.com/mauritsvanrees/c47e974e418c707a626200cb6561405b) that creates those files based on buildout configuration. Preferably you would run the script and then remove the version pins from the buildout config. Maybe buildout could parse the txt files instead. But it should not really be needed when we don't allow buildout to install Python packages.

One tricky problem is `z3c.autoinclude`. When a package uses `includeDependencies` in its `configure.zcml` file, this does not work well with pip-installed packages. See [issue 9](https://github.com/zopefoundation/z3c.autoinclude/issues/9). Core Plone does not need that part of `z3c.autoinclude`, but add-on packages might.

**Comment by gotcha** [comment]
See #502 where I have removed dependency on `setuptools.easy_install`.

**Comment by gotcha** [comment]
I opened #506 instead of #502

**Comment by gotcha** [comment]
I merged #506.
